runtime/gc_base/IndexableObjectAllocationModel.cpp

```diff
@@ -345,39 +465,120 @@ MM_IndexableObjectAllocationModel::doubleMapArraylets(MM_EnvironmentBase *env, J
+#endif /* J9VM_GC_DOUBLE_MAPPING_FOR_SPARSE_HEAP_ALLOCATION */
+		void *virtualLargeObjectHeapAddress = extensions->largeObjectVirtualMemory->allocateSparseFreeEntryAndMapToHeapObject(spine, byteAmount);
+		if (NULL != virtualLargeObjectHeapAddress) {
+			indexableObjectModel->setDataAddrForContiguous((J9IndexableObject *)spine, virtualLargeObjectHeapAddress);
+#if defined(J9VM_GC_DOUBLE_MAPPING_FOR_SPARSE_HEAP_ALLOCATION)
+			void *contiguousAddress = doubleMapArraylets(env, (J9Object *)spine, arrayletLeaveAddrs, firstLeafRegionDescriptor, virtualLargeObjectHeapAddress);
+			Assert_MM_true(virtualLargeObjectHeapAddress == contiguousAddress);
+			extensions->largeObjectVirtualMemory->recordDoubleMapIdentifierForData(virtualLargeObjectHeapAddress, &firstLeafRegionDescriptor->_arrayletDoublemapID);
+#endif /* defined(J9VM_GC_DOUBLE_MAPPING_FOR_SPARSE_HEAP_ALLOCATION) */
+		}
+		/* Free arraylet leaf addresses if dynamically allocated */
+		if (arrayletLeafCount > ARRAYLET_ALLOC_THRESHOLD) {
+			env->getForge()->free((void *)arrayletLeaveAddrs);
+		}
 	}
 
-	/* Number of arraylet leaves in the iterator must match the number of leaves calculated */
-	Assert_MM_true(arrayletLeafCount == count);
+	Trc_MM_getSparseAddressAndDecommitLeaves_Exit(env->getLanguageVMThread(), spine, (void *)bytesRemaining);
 
-	GC_SlotObject objectSlot(env->getOmrVM(), extensions->indexableObjectModel.getArrayoidPointer((J9IndexableObject *)objectPtr));
-	J9Object *firstLeafSlot = objectSlot.readReferenceFromSlot();
+	return spine;
+}
+#endif /* defined(J9VM_ENV_DATA64) */
 
-	MM_HeapRegionDescriptorVLHGC *firstLeafRegionDescriptor = (MM_HeapRegionDescriptorVLHGC *)heap->getHeapRegionManager()->tableDescriptorForAddress(firstLeafSlot);
+#if defined(J9VM_GC_ENABLE_DOUBLE_MAP)
+#if !((defined(LINUX) || defined(OSX)) && defined(J9VM_ENV_DATA64))
+/* Double map is only supported on OSX and LINUX 64 bit Systems for now */
+#error "Platform not supported by Double Map API"
+#endif /* !((defined(LINUX) || defined(OSX)) && defined(J9VM_ENV_DATA64)) */
+void *
+MM_IndexableObjectAllocationModel::doubleMapArraylets(MM_EnvironmentBase *env, J9Object *objectPtr, void **arrayletLeaveAddrs, MM_HeapRegionDescriptorVLHGC *firstLeafRegionDescriptor, void *preferredAddress)
+{
+	MM_GCExtensions *extensions = MM_GCExtensions::getExtensions(env);
+	MM_Heap *heap = extensions->getHeap();
+	UDATA arrayletLeafSize = env->getOmrVM()->_arrayletLeafSize;
+	UDATA arrayletLeafCount = MM_Math::roundToCeiling(arrayletLeafSize, _dataSize) / arrayletLeafSize;
 
 	/* Retrieve actual page size */
 	UDATA pageSize = heap->getPageSize();
 
 	/* For now we double map the entire region of all arraylet leaves. This might change in the future if hybrid regions are introduced. */
 	uintptr_t byteAmount = arrayletLeafSize * arrayletLeafCount;
 
+	Trc_MM_double_map_Entry(env->getLanguageVMThread(), (void *)_dataSize, (void *)byteAmount, (void *)objectPtr, (void *)arrayletLeafSize, arrayletLeafCount);
```

Comment:
This doesn't compile:
```
IndexableObjectAllocationModel.cpp:569:155: error: macro "Trc_MM_double_map_Entry" passed 6 arguments, but takes just 4
```

Reply:
Yes, please see the [PR](https://github.com/eclipse-openj9/openj9/pull/14667#issue-1161666297) description for the `depends on` PR listed there. This compilation issue, as well as other noted compilation issues, are resolved when the dependent OMR PR is applied.